_posts/2024-08-20-integration-for-kw-build.md

```diff
@@ -0,0 +1,517 @@
+    exit 1
+  else
+    return 0
+  fi
+}
+```
+
+### 5. Cancel Build Process
+
+To prevent the build process from continuing after monitoring, the script
+terminates all related build processes using `pstree` to find all subprocesses.
+
+```bash
+function cancel_build_processes()
+{
+  local pids_to_kill
+  local parent_pid
+  local parent_pids
+
+  # Using mapfile to populate parent_pids array
+  mapfile -t parent_pids < <(pgrep -f "$KW_BUILD_CMD" || true)
+
+  for parent_pid in "${parent_pids[@]}"; do
+    if [ -n "$parent_pid" ]; then
+      # Using read with IFS to populate pids_to_kill array
+      IFS=' ' read -r -a pids_to_kill <<< "$(pstree -p "$parent_pid" | grep -o '([0-9]\+)' | grep -o '[0-9]\+')"
+
+      printf "Cancelling PIDs: %s\n" "${pids_to_kill[@]}"
+      printf "%s\n" "${pids_to_kill[@]}" | xargs kill -9
+    fi
+  done
+}
+```
+
+### 6. Script Execution
+
+Finally, the script runs the kw build command in the background, monitors CPU
+usage, calculates the average, checks if it is within the error margin, and
+cancels processes at the end.
+
+```bash
+# Start the build command in the background
+eval "$KW_BUILD_CMD" &
+# Wait a short period to ensure the kw build process is running
+sleep 30
+# Monitor CPU usage while the process is running
+monitor_cpu_usage
+# Cancel the build processes and their subprocesses
+cancel_build_processes
+# Calculate the average CPU usage
+avg_cpu_usage=$(calculate_avg_cpu_usage)
+printf "Average CPU usage during build: %.2f%%\n" "$avg_cpu_usage"
+# Check if the average CPU usage is within the expected range
+check_cpu_usage "$avg_cpu_usage"
+# Clean up the CPU usage file
+rm $CPU_USAGE_FILE
+```
+
+### Summary of Testing kw build with –cpu-scaling option
+
+Returning to our cpu-scaling option test function:
+
+
+```bash
+function test_kernel_build_cpu_scaling_option()
+{
+  local build_status
+  local build_result
+  local build_status_log
+  local cpu_scaling_percentage=50
+
+  container_exec "$CONTAINER" "cd ${KERNEL_TREE_PATH_CONTAINER} && kw_build_cpu_scaling_monitor ${cpu_scaling_percentage} > /dev/null 2>&1"
+  assert_equals_helper "kw build --cpu-scaling 50 failed for ${CONTAINER}" "(${LINENO})" 0 "$?"
+}
+```
+
+#### Validating with assert_equals_helper
+
+The test runs inside a container, where the script monitors CPU usage while `kw
+build --cpu-scaling 50` is executed. The check_cpu_usage function compares the
+average CPU usage with the expected value and, based on this, returns 0
+(**success**) or 1 (**failure**). The result is then verified by
+`assert_equals_helper`, ensuring that the behavior is as expected.
+
+With this, we conclude the validation of the CPU scaling feature. If the
+check_cpu_usage function returns 0, the test is considered successful,
+validating that the CPU scaling functionality of kw build is working correctly.
+
+# Conclusion
+
+Integration testing for `kw build` is crucial to ensure the tool's robustness
```

Comment:
`kw build` is one of the core features of `kw`, so integration testing for it is crucial to ensure the tool's robustness"